### DISCUSSION TITLE
sftpFileStream Seek() limited to 2GB file

### DISCUSSION THREAD

**Discussion by NigelGeniq** [sshnet/SSH.NET#1445] [Q&A]
I have files on an SFTP server and am using the sftpFileStream object's Seek(offset, origin) function to move to points within the files before reading a block of data of predefined size from them.

// Seek to the start position of the current block
sftpStream.Seek(start, SeekOrigin.Begin);
// Read the current block into the buffer
int bytesRead = sftpStream.Read(buffer, 0, currentBlockSize);

This works fine for most files, however I have found that if the file on the SFTP server is larger than 2GB, any attempt to Seek() to a start point beyond 2GB throws an error. I assume this is because the underlying MemoryStream is itself limited to 2GB Max?

Is there an elegant way around this, without having to either insist on source files smaller than 2GB, or writing something to split them up first?

**Comment by Rob-Hague** [comment]
There is no MemoryStream in SftpFileStream, unless you are referring to somewhere else. In theory it should (i.e. is supposed to) work. What is the error+stack trace?

**Reply by NigelGeniq** [reply]
I couldn't tell you that now, as we have moved away from using it in order to find a resolution.

But my investigations / debugging yesterday did show something weird:

1. All I actually need to do is use "sftpStream.Read(buffer, 0, currentBlockSize);" repeatedly, and it steps through the whole file, with no 2GB problem.
2. "sftpStream.Seek(start, SeekOrigin.Begin);" caused a problem as soon as the value of "start" hit 2,147,483,648.
3. There was NO connection between what "sftpStream.Seek();" was doing and what "sftpStream.Read();" was doing - debugging and logging clearly showed that if I first used "sftpStream.Seek()" to move to a position in the stream of say 1.8GB, "sftpStream.Position" then showed it incrementing by "currentBlockSize" from there on every itteration. However, analysing the data being returned by "sftpStream.Read()" clearly showed that regardless of where I'd previosuly "seeked" to, it was reading from the start of the file, in chunks of "currentBlockSize". 
4. Using "sftpStream.Position = X" to preset the starting point before cycling through also had the same affeect - position was shown to increase from that point, but "Read()" actually still read from the start of the file.

Since the result of the above weirdness was actually what we need to happen, I was happy to just ditch the use of "sftpStream.Seek()".
